File: app/schemas/ap_runs.py
"""Request and response shapes for AP Orchestrator runs."""

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class RunRequest(BaseModel):
    """Start one Orchestrator run for one invoice."""

    invoice_ref: str = Field(min_length=1, max_length=64)
    run_id: str | None = Field(default=None, max_length=64)
    workflow_id: str | None = Field(default=None, max_length=128)
    trigger_source: str = Field(default="api", max_length=50)


class RunEventOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    seq: int
    event_type: str
    operator_name: str | None
    summary: str | None
    ts: datetime | None


class DecisionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    belnr: str
    lifnr: str | None
    ebeln: str | None
    bukrs: str | None
    currency: str | None
    amount: float | None
    amount_myr: float | None
    verdict: str
    reason_codes: list[str]
    money_protected: float
    spend_under_review: float
    policy_version_label: str | None
    confidence: float | None
    human_status: str | None


class PolicyEvaluationOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    policy_key: str
    policy_version: int
    threshold_value: Any | None
    observed_value: Any | None
    fired: bool
    outcome: str | None
    explanation: str | None


class RunSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    run_id: str
    invoice_ref: str | None
    workflow_run_id: str | None
    status: str
    trigger_source: str
    policy_version_label: str | None
    started_at: datetime | None
    finished_at: datetime | None
    duration_ms: int | None
    error: str | None


class RunDetail(RunSummary):
    events: list[RunEventOut] = []
    decision: DecisionOut | None = None
    policy_evaluations: list[PolicyEvaluationOut] = []
    workbench_item_id: int | None = None


class RunListResponse(BaseModel):
    runs: list[RunSummary]
    total: int

File: frontend/src/types/ap-runs.ts
/** Mirrors app/schemas/ap_runs.py. Keep in sync if that file changes. */

export interface RunRequest {
  invoice_ref: string
  run_id?: string
  workflow_id?: string
  trigger_source?: string
}

export interface DecisionOut {
  id: number
  belnr: string
  lifnr: string | null
  ebeln: string | null
  bukrs: string | null
  currency: string | null
  amount: number | null
  amount_myr: number | null
  verdict: string
  reason_codes: string[]
  money_protected: number
  spend_under_review: number
  policy_version_label: string | null
  confidence: number | null
  human_status: string | null
}

export interface PolicyEvaluationOut {
  policy_key: string
  policy_version: number
  threshold_value: unknown
  observed_value: unknown
  fired: boolean
  outcome: string | null
  explanation: string | null
}

export interface RunEventOut {
  seq: number
  event_type: string
  operator_name: string | null
  summary: string | null
  ts: string | null
}

export interface RunDetail {
  run_id: string
  invoice_ref: string | null
  workflow_run_id: string | null
  status: string
  trigger_source: string
  policy_version_label: string | null
  started_at: string | null
  finished_at: string | null
  duration_ms: number | null
  error: string | null
  events: RunEventOut[]
  decision: DecisionOut | null
  policy_evaluations: PolicyEvaluationOut[]
  workbench_item_id: number | null
}
